NetKet/Machine/activation_layer.hpp

```diff
@@ -0,0 +1,159 @@
+  using VectorConstRefType = typename AbstractLayer<T>::VectorConstRefType;
+
+  A activation_;  // activation
+  int size_;      // size_ = input size = output size
+ public:
+  using StateType = typename AbstractLayer<T>::StateType;
+  using LookupType = typename AbstractLayer<T>::LookupType;
+
+  /// Constructor
+  Activation(const int input_size) : activation_(), size_(input_size) {
+    Init();
+  }
+
+  void Init() {
+    std::string buffer = "";
+    InfoMessage(buffer) << activation_.name << " Activation Layer " << size_
+                        << " --> " << size_ << std::endl;
+  }
+
+  void to_json(json &pars) const override {
+    json layerpar;
+    layerpar["Name"] = "Activation";
+    layerpar["Size"] = size_;
+
+    pars["Machine"]["Layers"].push_back(layerpar);
+  }
+
+  void from_json(const json & /*pars*/) override {}
+
+  void InitRandomPars(int /*seed*/, double /*sigma*/) override {}
+
+  int Npar() const override { return 0; }
+
+  int Ninput() const override { return size_; }
+
+  int Noutput() const override { return size_; }
+
+  void GetParameters(VectorRefType /*pars*/, int /*start_idx*/) const override {
+  }
+
+  void SetParameters(VectorConstRefType /*pars*/, int /*start_idx*/) override {}
+
+  void InitLookup(const VectorType &v, LookupType &lt,
+                  VectorType &output) override {
+    lt.resize(1);
+    lt[0].resize(size_);
+
+    Forward(v, lt, output);
+  }
+
+  void UpdateLookup(const VectorType & /*input*/,
+                    const std::vector<int> &input_changes,
+                    const VectorType &new_input, LookupType &theta,
+                    const VectorType & /*output*/,
+                    std::vector<int> &output_changes,
+                    VectorType &new_output) override {
+    const int num_of_changes = input_changes.size();
+    if (num_of_changes == size_) {
+      output_changes.resize(size_);
+      new_output.resize(size_);
+      theta[0] = new_input;
+      activation_.operator()(new_input, new_output);
+    } else if (num_of_changes > 0) {
+      output_changes = input_changes;
+      new_output.resize(num_of_changes);
+      for (int i = 0; i < num_of_changes; ++i) {
+        theta[0](input_changes[i]) = new_input(i);
+      }
+      activation_.operator()(new_input, new_output);
+    } else {
+      output_changes.resize(0);
+      new_output.resize(0);
+    }
+  }
+
+  void UpdateLookup(const Eigen::VectorXd & /*input*/,
+                    const std::vector<int> &tochange,
+                    const std::vector<double> &newconf, LookupType &theta,
+                    const VectorType & /*output*/,
+                    std::vector<int> &output_changes,
+                    VectorType &new_output) override {
+    const int num_of_changes = tochange.size();
+    if (num_of_changes > 0) {
+      output_changes = tochange;
+      new_output.resize(num_of_changes);
+      for (int i = 0; i < num_of_changes; ++i) {
+        theta[0](tochange[i]) = newconf[i];
+      }
```

Comment:
Is this a "light-cone" like implementation that you didn't have for the old activations?

Reply:
yea but honestly at the moment the light cones are not handled very well. Now only the first layer will take advantage of light cones